src/List.h

```diff
@@ -20,67 +20,353 @@
 //	Entries must be either a pointer to the data or nonzero data with
 //	sizeof(data) <= sizeof(void*).
 
+#include <stdarg.h>
 #include <initializer_list>
+#include <iterator>
 #include <utility>
-#include <stdarg.h>
 #include "util.h"
 
-typedef void* ent;
 typedef int (*list_cmp_func)(const void* v1, const void* v2);
 
-class BaseList {
+template<typename T>
+class ListIterator
+	{
+public:
+	ListIterator(T* entries, int offset, int num_entries) :
+		entries(entries), offset(offset), num_entries(num_entries), endptr() {}
+	bool operator==(const ListIterator& rhs) { return entries == rhs.entries && offset == rhs.offset; }
+	bool operator!=(const ListIterator& rhs) { return entries != rhs.entries || offset != rhs.offset; }
+	ListIterator & operator++() { offset++; return *this; }
+	ListIterator operator++(int) { auto t = *this; offset++; return t; }
+	ListIterator & operator--() { offset--; return *this; }
+	ListIterator operator--(int) { auto t = *this; offset--; return t; }
+	std::ptrdiff_t operator-(ListIterator const& sibling) const { return offset - sibling.offset; }
+	ListIterator & operator+=(int amount) { offset += amount; return *this; }
+	ListIterator & operator-=(int amount) { offset -= amount; return *this; }
+	bool operator<(ListIterator const&sibling) const { return offset < sibling.offset;}
+	bool operator<=(ListIterator const&sibling) const { return offset <= sibling.offset; }
+	bool operator>(ListIterator const&sibling) const { return offset > sibling.offset; }
+	bool operator>=(ListIterator const&sibling) const { return offset >= sibling.offset; }
+	T& operator[](int index)
+		{
+		if (index < num_entries)
+			return entries[index];
+		else
+			return endptr;
+		}
+	T& operator*()
+		{
+		if ( offset < num_entries )
+			return entries[offset];
+		else
+			return endptr;
+		}
+
+private:
+	T* const entries;
+	int offset;
+	int num_entries;
+	T endptr; // let this get set to some random value on purpose. It's only used
+			  // for the operator[] and operator* cases where you pass something
+			  // off the end of the collection, which is undefined behavior anyways.
+	};
+
+
+namespace std {
+	template<typename T>
+	class iterator_traits<ListIterator<T> >
+	{
+	public:
+		using difference_type = std::ptrdiff_t;
+		using size_type = std::size_t;
+		using value_type = T;
+		using pointer = T;
+		using reference = T&;
+		using iterator_category = std::random_access_iterator_tag;
+	};
+}
+
+
+template<typename T>
+class List {
 public:
-	void clear();		// remove all entries
+
```

Comment:
This is missing `using value_type = T;` which will make it play nicer with things out of the STL.

Reply:
It's there, just down by where the iterator traits are defined.

Reply:
nice :) sorry about that.